File: 프로그래머스/lv2/1844. 게임 맵 최단거리/게임 맵 최단거리.js
function solution(maps) {
    let dx = [1,-1,0,0];
    let dy = [0,0,1,-1];
    
    let n = maps[0].length;
    let m = maps.length;
    
    let queue = [[0,0]];
    maps[0][0] = 0;
    let answer = 1;
    
    while(queue.length){
        let size = queue.length;
        for(let i=0; i<size; i++){
            let [x,y] = queue.shift();
            for(let j=0; j<4; j++){
                let nx = x + dx[j];
                let ny = y + dy[j];
                if(nx >= 0 && ny >= 0 && nx < n && ny < m && maps[ny][nx] === 1){
                    if(nx === n-1 && ny === m-1){
                     return ++answer;
                    }
                    queue.push([nx,ny]);
                    maps[ny][nx] = 0;
                }
            }
        }
        answer++;
    }
    return -1;
}

File: 프로그래머스/2/1844. 게임 맵 최단거리/게임 맵 최단거리.java
import java.util.ArrayDeque;

class Solution {
    public int solution(int[][] maps) {
        int n = maps[0].length; //5
        int m = maps.length; // 5
        int[][] move = new int[][]{{1,0},{-1,0},{0,1},{0,-1}};
        int answer = 1;
        ArrayDeque<int[]> queue = new ArrayDeque<>();
        queue.offer(new int[]{0, 0});
        maps[0][0] = 0;

        while (!queue.isEmpty()) {
            int queueSize = queue.size();
            for (int i = 0; i < queueSize; i++) {
                int[] coor = queue.poll();

                int y = coor[0];
                int x = coor[1];

                for (int j = 0; j < move.length; j++) {
                    int mx = x + move[j][0];  // 1 -1  0  0
                    int my = y + move[j][1];  // 0  0  1 -1

                    if (mx >= 0 && mx < n && my >= 0 && my < m && maps[my][mx] == 1) {
                        if (mx == n - 1 && my == m - 1) {
                            return ++answer;
                        }
                        queue.offer(new int[]{my, mx});
                        maps[my][mx] = 0;
                    }
                }
            }
            answer++;
        }

        return -1;
    }
}